shared/domains/chelonia/chelonia.js

```diff
@@ -792,48 +789,113 @@ export default (sbp('sbp/selectors/register', {
-        if (has(this.removeCount, contractID)) {
-          if (this.removeCount[contractID] > 1) {
-            this.removeCount[contractID] -= 1
-          } else {
-            delete this.removeCount[contractID]
+      return sbp('chelonia/private/queueEvent', contractID, () => {
+        const rootState = sbp(this.config.stateSelector)
+        const fkContractIDs = Array.from(new Set(Object.values(rootState[contractID]?._vm?.authorizedKeys ?? {}).filter((k) => {
+          return !!(k: any).foreignKey
+        }).map((k) => {
+          try {
+            const fkUrl = new URL((k: any).foreignKey)
+            return fkUrl.pathname
+          } catch {
+            return undefined
           }
-        }
-        if (!rootState.contracts[contractID].pendingRemove) {
-          return undefined
-        }
-      }
+        }).filter(Boolean)))
 
-      if (this.removeCount[contractID] >= 1) {
-        rootState.contracts[contractID].pendingRemove = true
-        return undefined
-      }
+        sbp('chelonia/private/removeImmediately', contractID)
 
-      return sbp('chelonia/private/queueEvent', contractID, [
-        'chelonia/private/removeImmediately', contractID
-      ])
+        if (fkContractIDs.length) {
+          // Attempt to release all contracts that are being monitored for
+          // foreign keys
+          sbp('chelonia/contract/release', fkContractIDs, { try: true }).catch((e) => {
+            console.error('[chelonia] Error attempting to release foreign key contracts', e)
+          })
+        }
+      })
     }))
   },
+  'chelonia/contract/retain': function (contractIDs: string | string[], params?: { ephemeral?: boolean}): Promise<*> {
+    const listOfIds = typeof contractIDs === 'string' ? [contractIDs] : contractIDs
+    if (listOfIds.length === 0) return Promise.resolve()
+    if (!params?.ephemeral) {
+      const rootState = sbp(this.config.stateSelector)
+      listOfIds.forEach((id) => {
+        if (!has(rootState.contracts, id)) {
+          this.config.reactiveSet(rootState.contracts, id, Object.create(null))
+        }
+        this.config.reactiveSet(rootState.contracts[id], 'references', (rootState.contracts[id].references ?? 0) + 1)
+      })
+    } else {
+      listOfIds.forEach((id) => {
+        if (!has(this.ephemeralReferenceCount, id)) {
+          this.ephemeralReferenceCount[id] = 1
+        } else {
+          this.ephemeralReferenceCount[id] = this.ephemeralReferenceCount[id] + 1
+        }
+      })
+    }
+    return sbp('chelonia/contract/sync', listOfIds)
+  },
+  // the `try` parameter does not affect (ephemeral or persistent) reference
+  // counts, but rather removes a contract if the reference count is zero
+  // and the contract isn't being monitored for foreign keys. This parameter
+  // is meant mostly for internal chelonia use, so that removing or releasing
+  // a contract can also remove other contracts that this first contract
+  // was monitoring.
```

Comment:
Could you add the following elaboration on the `try` parameter that explains why it exists:

> The `try` parameter is an internal parameter (not intended for use by developers), that helps Chelonia remove foreign contracts that are no longer being watched.
> The reason why we have it, and cannot simply use reference counting to do this automatically, is because we have two different types of reference counting (persistent and ephemeral). And therefore, if, for example, there was a foreign contract that was monitored for an ephemeral reason at one point, and a persistent reason at a different point, then because ephemeral reasons don't get saved like persistent reasons, it would be difficult to know when to unsubscribe from it.

TBD: tomorrow come up with a concrete scenario that proves that the above is a true statement.

Reply:
```ts
sbp('chelonia/contract/retain', 'A', { ephemeral: true })
// Will also call sbp('chelonia/private/in/syncContractAndWatchKeys', 'B')
// Now 'A' has ephemeral count set to 1 and no non-ephemeral count
// Say that 'B' also has ephemeral count set to 1 and no non-ephemeral count
// (because we use `ephemeral` from the call to 'A')

sbp('chelonia/contract/retain', 'A', { ephemeral: false })
// Now 'A' has ephemeral count set to 1 and non-ephemeral count set to 1
// What do we do with 'B'?

sbp('chelonia/contract/release', 'A', { ephemeral: true })
// Now 'A' has no ephemeral count set and non-ephemeral count set to 1
// What do we do with 'B'?
// If the page is refreshed, 'B' won't have any reference count and will be
// released, which is wrong.

sbp('chelonia/contract/release', 'A', { ephemeral: false })
// Now we remove 'A'
// How do we know whether to use `{ ephemeral: true }`? Using the wrong one will
// result in an exception.

```